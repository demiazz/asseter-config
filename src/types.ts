/* tslint:disable: interface-name */

type JSONScalar = boolean | null | number | string | void;

interface JSONArray extends Array<JSONValue> {}

interface JSONObject extends Record<string, JSONValue> {}

export type JSONValue = JSONScalar | JSONArray | JSONObject;

type PackageManager = "npm" | "yarn";

type ProviderOptions = JSONObject;

type ProviderEnvironmentOptions = Record<string, ProviderOptions>;

interface Provider extends JSONObject {
  environment?: ProviderEnvironmentOptions;
  options: ProviderOptions;
  type: string;
}

interface Providers extends Record<string, Provider> {
  default: Provider;
}

export interface Configuration extends JSONObject {
  defaultEnvironment: string;
  environmentVariable: string;
  packageManager: PackageManager;
  providers: Providers;
}
