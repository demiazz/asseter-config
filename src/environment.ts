import camelCase from "camel-case";
import constantCase from "constant-case";

import { JSONScalar } from "./types";

const isOctal = (value: string): boolean => /^[-+]?0\d+$/.test(value);

const isHexadecimal = (value: string): boolean =>
  /^[-+]?0(x|X)[0-9A-Fa-f]+$/.test(value);

const isDecimal = (value: string): boolean => /^[-+]?\d+$/.test(value);

const isFloat = (value: string): boolean =>
  /^[-+]?\d*\.?\d+([eE][-+]?\d+)?$/.test(value);

export const parse = (value: string): boolean | null | number | string => {
  if (value === "true") {
    return true;
  }

  if (value === "false") {
    return false;
  }

  if (value === "null") {
    return null;
  }

  if (isOctal(value)) {
    return parseInt(value, 8);
  }

  if (isHexadecimal(value)) {
    return parseInt(value, 16);
  }

  if (isDecimal(value)) {
    return parseInt(value, 10);
  }

  if (isFloat(value)) {
    return parseFloat(value);
  }

  return value;
};

export const extractFromEnvironment = (
  namespace: string
): Record<string, JSONScalar> => {
  const prefix = `ASSETER_${constantCase(namespace)}_`;
  const variables: Record<string, JSONScalar> = {};

  for (const variableName of Object.keys(process.env)) {
    if (!variableName.startsWith(prefix)) {
      continue;
    }

    const value = process.env[variableName];

    if (value == null) {
      continue;
    }

    const name = camelCase(variableName.substr(prefix.length));

    variables[name] = parse(value);
  }

  return variables;
};
