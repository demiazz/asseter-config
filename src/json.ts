export type JSONScalar = boolean | null | number | string | void;

export interface JSONArray extends Array<JSONValue> {}

export interface JSONObject extends Record<string, JSONValue> {}

export type JSONValue = JSONScalar | JSONArray | JSONObject;
