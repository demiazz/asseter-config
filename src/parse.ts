const isOctal = (value: string): boolean => /^\-?0\d+$/.test(value);

const isHexadecimal = (value: string): boolean =>
  /^\-?0(x|X)[0-9A-Fa-f]+$/.test(value);

const isDecimal = (value: string): boolean => /^\-?\d+$/.test(value);

const isFloat = (value: string): boolean =>
  /^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$/.test(value);

export const parse = (value: string): boolean | null | number | string => {
  if (value === "true") {
    return true;
  } else if (value === "false") {
    return false;
  } else if (value === "null") {
    return null;
  } else if (isOctal(value)) {
    return parseInt(value, 8);
  } else if (isHexadecimal(value)) {
    return parseInt(value, 16);
  } else if (isDecimal(value)) {
    return parseInt(value, 10);
  } else if (isFloat(value)) {
    return parseFloat(value);
  }

  return value;
};
