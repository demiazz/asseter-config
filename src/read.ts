import Ajv, { ValidateFunction } from "ajv";
import { readFileSync } from "fs";
import { safeLoad as parseYAML } from "js-yaml";
import { extname } from "path";
import { parse as parseTOML } from "toml";

import { JSONValue } from './types';

type Parser = (content: string) => JSONValue;

const getParser = (fileName: string): Parser | void => {
  switch (extname(fileName)) {
    case ".json":
      return JSON.parse;
    case ".yml":
    case ".yaml":
      return parseYAML;
    case ".toml":
      return parseTOML;
    default:
      return undefined;
  }
};

export const read = (fileName: string): JSONValue => {
  const parse = getParser(fileName);

  if (!parse) {
    throw new Error("Supported only JSON, YAML and TOML file types");
  }

  const content = readFileSync(fileName, { encoding: "utf8" });

  return parse(content);
};

export const readSchema = (definitionPath: string): ValidateFunction => {
  const definition = read(definitionPath);
  const compiler = new Ajv({
    allErrors: true,
    jsonPointers: false,
  });

  return compiler.compile(definition as object);
}
