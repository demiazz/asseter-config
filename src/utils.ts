export const reduceKeys = <U>(
  object: Record<string, any>,
  callback: (
    previousValue: U,
    currentValue: string,
    currentIndex: number,
    array: string[]
  ) => U,
  initialValue: U
) => Object.keys(object).reduce<U>(callback, initialValue);
