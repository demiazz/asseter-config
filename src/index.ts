import { read, readSchema } from "./read";
import { ValidationError } from "./validate";

export const validate = (
  schemaFileName: string,
  configurationFileName: string
): void => {
  const schema = readSchema(schemaFileName);
  const configuration = read(configurationFileName);

  if (!schema(configuration)) {
    throw new ValidationError(
      (schema.errors || []).map(({ dataPath, message = "" }) => ({
        message,
        path: dataPath
      }))
    );
  }
};
