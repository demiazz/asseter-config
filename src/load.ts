import { resolve } from "path";

import { read } from "./read";
import { JSONValue, RawConfiguration } from "./types";
import { Errors, validate, ValidationError } from "./validate";

const schemaPath = resolve(__dirname, "../schema.json");

const isRawConfiguration = (
  _: JSONValue,
  errors: Errors
): _ is RawConfiguration => {
  return errors.length === 0;
};

export const load = (filePath: string): RawConfiguration => {
  const data: JSONValue = read(filePath);
  const rawConfigurationErrors: Errors = validate(schemaPath, data);

  if (!isRawConfiguration(data, rawConfigurationErrors)) {
    throw new ValidationError(rawConfigurationErrors);
  }

  const rawConfiguration: RawConfiguration = data;

  return rawConfiguration;
};
