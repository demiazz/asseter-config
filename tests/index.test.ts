import { validate } from "../src";

import { getFixturePath } from "./helpers";

describe("validate", () => {
  describe("when configuration is valid", () => {
    it("doesn't throw an error", () => {
      const schemaPath = getFixturePath("index/schema.json");
      const configurationPath = getFixturePath("index/valid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).not.toThrowError();
    });
  });

  describe("when configuration is invalid", () => {
    it("throws an error", () => {
      const schemaPath = getFixturePath("index/schema.json");
      const configurationPath = getFixturePath("index/invalid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).toThrowError();
    });

    it("throws an error with all errors information", () => {
      const schemaPath = getFixturePath("index/schema.json");
      const configurationPath = getFixturePath("index/invalid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
