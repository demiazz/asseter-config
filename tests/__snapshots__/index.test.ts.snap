// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate when configuration is invalid throws an error with all errors information 1`] = `
"Invalid configuration:
  root should NOT have additional properties
  root.isValid should be equal to one of the allowed values"
`;
